src/coreclr/jit/valuenum.cpp

```diff
@@ -5403,6 +5405,21 @@ ValueNum ValueNumStore::VNForLoadStoreBitCast(ValueNum value, var_types indType,
 {
     var_types typeOfValue = TypeOfVN(value);
 
+    // BitCast<TYP_REF>(TypeHandleToRuntimeTypeHandle(clsHandle)) => frozen RuntimeType handle
+    //
+    VNFuncApp funcApp;
+    if ((indType == TYP_REF) && GetVNFunc(value, &funcApp) && (funcApp.m_func == VNF_TypeHandleToRuntimeTypeHandle) &&
+        IsVNHandle(funcApp.m_args[0]) && (GetHandleFlags(funcApp.m_args[0]) == GTF_ICON_CLASS_HDL))
+    {
+        CORINFO_CLASS_HANDLE  cls     = (CORINFO_CLASS_HANDLE)ConstantValue<ssize_t>(funcApp.m_args[0]);
+        CORINFO_OBJECT_HANDLE typeObj = m_pComp->info.compCompHnd->getRuntimeTypePointer(cls);
+        if (typeObj != nullptr)
+        {
+            m_pComp->setMethodHasFrozenObjects();
+            return VNForHandle((ssize_t)typeObj, GTF_ICON_OBJ_HDL);
+        }
+    }
```

Comment:
`VNForLoadStoreBitCast` is not the right place to insert auxiliary optimizations. `VNForBitCast` would be the place to do it.

Reply:
Moved there